File: src/main/java/dsl/TypedBranchTargetBuilder.java
package dsl;

import java.util.function.BiConsumer;
import java.util.function.Function;

public class TypedBranchTargetBuilder<S, T, V, U> {
    private final TypedBranchActionBuilder<S, T> parent;
    private final Function<S, V> sourceAccessor;
    private final BiConsumer<T, U> targetSetter;
    private final Function<V, Object> transformation;

    TypedBranchTargetBuilder(TypedBranchActionBuilder<S, T> parent,
                             Function<S, V> sourceAccessor,
                             BiConsumer<T, U> targetSetter,
                             Function<V, Object> transformation) {
        this.parent = parent;
        this.sourceAccessor = sourceAccessor;
        this.targetSetter = targetSetter;
        this.transformation = transformation;
    }

    public TypedBranchActionBuilder<S, T> end() {
        // Use the properly typed mapping rule
        parent.addAction(new TypedBranchMapping<>(sourceAccessor, targetSetter, transformation, null));
        return parent;
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/typed/TypedBranchTargetBuilder.kt
package xyz.mahmoudahmed.dsl.typed

class TypedBranchTargetBuilder<S, T, V, U>(
    private val parent: TypedBranchActionBuilder<S, T>,
    private val sourceAccessor: (S) -> V,
    private val targetSetter: (T, U) -> Unit,
    private val transformation: ((V) -> Any)?
) {
    fun end(): TypedBranchActionBuilder<S, T> {
        parent.addAction(TypedBranchMapping(sourceAccessor, targetSetter, transformation, null))
        return parent
    }
}